mne_connectivity/decoding/tests/test_decomposition.py

```diff
@@ -0,0 +1,672 @@
+            info=epochs.info,
+            method=method,
+            indices=indices,
+            mode=mode,
+            fmin=fmin,
+            fmax=fmax,
+            cwt_freqs=cwt_freqs,
+            cwt_n_cycles=cwt_n_cycles,
+            rank=(0, 1),
+        )
+    with pytest.raises(
+        ValueError,
+        match=(
+            "at least one entry in `rank` is greater than the number of seed/target "
+            "channels in `indices`"
+        ),
+    ):
+        CoherencyDecomposition(
+            info=epochs.info,
+            method=method,
+            indices=indices,
+            mode=mode,
+            fmin=fmin,
+            fmax=fmax,
+            cwt_freqs=cwt_freqs,
+            cwt_n_cycles=cwt_n_cycles,
+            rank=(n_seeds + 1, n_targets),
+        )
+    with pytest.raises(
+        ValueError,
+        match=(
+            "at least one entry in `rank` is greater than the number of seed/target "
+            "channels in `indices`"
+        ),
+    ):
+        CoherencyDecomposition(
+            info=epochs.info,
+            method=method,
+            indices=indices,
+            mode=mode,
+            fmin=fmin,
+            fmax=fmax,
+            cwt_freqs=cwt_freqs,
+            cwt_n_cycles=cwt_n_cycles,
+            rank=(n_seeds, n_targets + 1),
+        )
+
+    # Test n_jobs
+    with pytest.raises(TypeError, match="`n_jobs` must be an instance of int"):
+        CoherencyDecomposition(
+            info=epochs.info,
+            method=method,
+            indices=indices,
+            mode=mode,
+            fmin=fmin,
+            fmax=fmax,
+            cwt_freqs=cwt_freqs,
+            cwt_n_cycles=cwt_n_cycles,
+            n_jobs="1",
+        )
+
+    # Test verbose
+    with pytest.raises(
+        TypeError, match="`verbose` must be an instance of bool, str, int, or None"
+    ):
+        CoherencyDecomposition(
+            info=epochs.info,
+            method=method,
+            indices=indices,
+            mode=mode,
+            fmin=fmin,
+            fmax=fmax,
+            cwt_freqs=cwt_freqs,
+            cwt_n_cycles=cwt_n_cycles,
+            verbose=[True],
+        )
+
+    decomp_class = CoherencyDecomposition(
+        info=epochs.info,
+        method=method,
+        indices=indices,
+        mode=mode,
+        fmin=fmin,
+        fmax=fmax,
+        cwt_freqs=cwt_freqs,
+        cwt_n_cycles=cwt_n_cycles,
+    )
+
+    # TEST BAD FITTING
+    # Test input data
+    with pytest.raises(TypeError, match="`X` must be an instance of NumPy array"):
+        decomp_class.fit(X=epochs.get_data().tolist())
+    with pytest.raises(ValueError, match="Invalid value for the '`X.ndim`' parameter"):
+        decomp_class.fit(X=epochs.get_data()[0])
+    with pytest.raises(ValueError, match="`X` does not match Info"):
+        decomp_class.fit(X=epochs.get_data()[:, :-1])
+    # XXX: Add test for rank of X being <= n_components when n_components can be > 1
```

Comment:
You can quickly fake a rank-deficient `X` by doing a PCA to remove ranks:

```
from sklearn.decomposition import PCA
X = epochs.get_data().reshape(len(epochs), -1).T
p = PCA(n_components=10).fit(1)  # some n_components < n_channels
X = p.transform(X)
X = p.inverse_transform(X)
X = X.T.reshape(epochs.get_data().shape)
```

Reply:
This is along the lines of how I would test this, however for now whilst `n_components` can only be 1 I am manually setting this:
https://github.com/mne-tools/mne-connectivity/blob/bdfb626382a8172fb5afad2c79ddafeb0e0e47fb/mne_connectivity/decoding/decomposition.py#L223

So even with rank-deficient data, the rank will always be >= `n_components`.

In the PR when support for >1 components is added then I would be able to clean up a few of these outstanding tests/missing coverage.

Reply:
Updated to manually set `n_components` so these tests can be run.